### PR TITLE
Add DevAddrPrefix to DevAddr generation function 

### LINKED ISSUE #130: Generate DevAddr from a prefix
**Summary:** 

NS should generate DevAddr from a supplied prefix with a fallback to NetID

**Why do we need this?**

@johanstokking wants this(https://github.com/TheThingsNetwork/lorawan-stack/pull/101#discussion_r255193262)

**What is already there? What do you see now?**

Generation of DevAddr from NetID only

**What is missing? What do you want to see?**

DevAddr from a supplied prefix with a fallback to NetID

**How do you propose to implement this?**

Store the prefix and use it to generate the DevAddr

### PR DESCRIPTION
**Pull request by furtiman**
**Summary:**

Closes #130 

**Changes:**

- Add random prefix choice from the network server slice of devAddrPrefixes
- Apply prefix to already generated DevAddr
- Make the network server create a prefix from NetID in case []types.DevAddrPrefixes is empty
- Add internal tests (grpc_gsns_internal_test.go)

**Notes for Reviewers:**

#222 is closely related. Should be considered next.

### CODE REVIEW

File: pkg/networkserver/config.go [hunk 1]
```diff
@@ -26,7 +26,8 @@ import (
 type Config struct {
 	Devices             DeviceRegistry         `name:"-"`
 	DownlinkTasks       DownlinkTaskQueue      `name:"-"`
-	NetID               types.NetID            `name:"net_id" description:"NetID of this Network Server"`
+	NetID               types.NetID            `name:"net-id" description:"NetID of this Network Server"`
+	DevAddrPrefixes     []types.DevAddrPrefix  `name:"dev-addr-prefixes" description:"DevAddrPrefixes of this Network Server"`
```
Comment (johanstokking) [inline pkg/networkserver/config.go]: Let's spell this out; `Device address prefixes`, in contrary to `NetID` which is a LoRaWAN term

File: pkg/networkserver/grpc_gsns.go [hunk 2]
```diff
@@ -781,11 +781,15 @@ func (ns *NetworkServer) handleUplink(ctx context.Context, up *ttnpb.UplinkMessa
 
 // newDevAddr generates a DevAddr for specified EndDevice.
 func (ns *NetworkServer) newDevAddr(context.Context, *ttnpb.EndDevice) types.DevAddr {
-	// TODO: Allow generating DevAddr from a prefix. (https://github.com/TheThingsNetwork/lorawan-stack/issues/130)
+	// TODO: Add tests
```
Comment (johanstokking) [inline pkg/networkserver/grpc_gsns.go]: Add testing

File: pkg/networkserver/grpc_gsns.go [hunk 3]
```diff
@@ -781,11 +781,15 @@ func (ns *NetworkServer) handleUplink(ctx context.Context, up *ttnpb.UplinkMessa
...
 	nwkAddr := make([]byte, types.NwkAddrLength(ns.netID))
 	random.Read(nwkAddr)
 	nwkAddr[0] &= 0xff >> (8 - types.NwkAddrBits(ns.netID)%8)
```
Comment (johanstokking) [inline pkg/networkserver/grpc_gsns.go]: We don't need this anymore

File: pkg/networkserver/grpc_gsns.go [hunk 4]
```diff
@@ -781,11 +781,15 @@ func (ns *NetworkServer) handleUplink(ctx context.Context, up *ttnpb.UplinkMessa
...
+
+	prefix := ns.devAddrPrefixes[random.Intn(len(ns.devAddrPrefixes))]
+
 	devAddr, err := types.NewDevAddr(ns.netID, nwkAddr)
```
Comment (johanstokking) [inline pkg/networkserver/grpc_gsns.go]: We don't need the NetID here. We don't need any reference to NetID in this function anymore.

File: pkg/networkserver/networkserver.go [hunk 5]
```diff
@@ -113,6 +113,8 @@ type NetworkServer struct {
 	devices DeviceRegistry
 
 	netID types.NetID
+	// TODO: Read from configuration
+	devAddrPrefixes []types.DevAddrPrefix
```
Comment (johanstokking) [inline pkg/networkserver/networkserver.go]: Take this into account too

File: pkg/networkserver/grpc_gsns.go [hunk 6]
```diff
@@ -781,11 +781,15 @@ func (ns *NetworkServer) handleUplink(ctx context.Context, up *ttnpb.UplinkMessa
 
 // newDevAddr generates a DevAddr for specified EndDevice.
 func (ns *NetworkServer) newDevAddr(context.Context, *ttnpb.EndDevice) types.DevAddr {
-	// TODO: Allow generating DevAddr from a prefix. (https://github.com/TheThingsNetwork/lorawan-stack/issues/130)
+	// TODO: Add tests
 	nwkAddr := make([]byte, types.NwkAddrLength(ns.netID))
 	random.Read(nwkAddr)
 	nwkAddr[0] &= 0xff >> (8 - types.NwkAddrBits(ns.netID)%8)
+
+	prefix := ns.devAddrPrefixes[random.Intn(len(ns.devAddrPrefixes))]
```
Comment (johanstokking) [inline pkg/networkserver/grpc_gsns.go]: We need to check configuration in `pkg/networkserver.New()` that there are in fact device address prefixes.

Or, if there are none, add one empty based on the NetID.

**Comment by furtiman** [comment]
@htdvisser Can you take a look?

File: pkg/networkserver/networkserver.go [hunk 7]
```diff
@@ -198,7 +198,7 @@ func New(c *component.Component, conf *Config, opts ...Option) (*NetworkServer,
 	if len(devAddrPrefixes) == 0 {
 		devAddr, err := types.NewDevAddr(conf.NetID, nil)
 		if err != nil {
-			return nil, err
+			panic(errors.New("failed to create new DevAddr").WithCause(err))
```
Comment (johanstokking) [inline pkg/networkserver/networkserver.go]: Why?

This may come from config; this is not an implementation issue necessarily. Just return an error

Comment (furtiman) [reply pkg/networkserver/networkserver.go]: removed

**Comment by johanstokking** [comment]
@htdvisser we already have a context available; signature is;

```go
 `(*NetworkServer).newDevAddr(context.Context, *ttnpb.EndDevice)`
```

So there we can check if context provides a custom prefix, falling back to instance level prefix if not. The idea is that we keep the functionality in this PR and only add another way to provide a `DevAddrPrefix` there.

**Comment by johanstokking** [comment]
@furtiman can you drop the last two commits to clean up the commit trail?

**Comment by furtiman** [comment]
@johanstokking done.

**Pull request merged**